test/tests_for_integration/auth_controller.py

```diff
@@ -0,0 +1,160 @@
+            raise TestException("root_temp_dir is None")
+
+        if bool(mongo_user) ^ bool(mongo_pwd):  # xor
+            raise TestException(
+                "Neither or both of mongo_user and mongo_pwd is required"
+            )
+
+        jars_dir = jars_dir.resolve()
+        class_path = self._get_class_path(jars_dir)
+
+        # make temp dirs
+        root_temp_dir = root_temp_dir.absolute()
+        os.makedirs(root_temp_dir, exist_ok=True)
+        self.temp_dir = Path(
+            tempfile.mkdtemp(prefix="AuthController-", dir=str(root_temp_dir))
+        )
+
+        self.port = test_utils.find_free_port()
+
+        template_dir = self.temp_dir.joinpath("templates")
+        self._install_templates(jars_dir, template_dir)
+
+        command = [
+            "java",
+            "-classpath",
+            class_path,
+            "-DAUTH2_TEST_MONGOHOST=" + mongo_host,
+            "-DAUTH2_TEST_MONGODB=" + mongo_db,
+            "-DAUTH2_TEST_TEMPLATE_DIR=" + str(template_dir),
+            _AUTH_CLASS,
+            str(self.port),
+        ]
+        if mongo_user:
+            command.insert(5, "-DAUTH2_TEST_MONGOPWD=" + mongo_pwd)
+            command.insert(5, "-DAUTH2_TEST_MONGOUSER=" + mongo_user)
+
+        self._outfile = open(self.temp_dir.joinpath("auth.log"), "w")
+
+        self._proc = subprocess.Popen(
+            command, stdout=self._outfile, stderr=subprocess.STDOUT
+        )
+
+        for count in range(40):
+            err = None
+            time.sleep(1)  # wait for server to start
+            try:
+                res = requests.get(
+                    f"http://localhost:{self.port}",
+                    headers={"accept": "application/json"},
+                )
+                if res.ok:
+                    self.version = res.json()["version"]
+                    break
+                err = TestException(res.text)
+            except requests.exceptions.ConnectionError as e:
+                err = TestException(e.args[0])
+                err.__cause__ = e
+        if err:
+            raise err
+        self.startup_count = count + 1
+
+    def destroy(self, delete_temp_files: bool = True):
+        """
+        Shut down the server and optionally delete any files generated.
+
+        :param delete_temp_files: if true, delete all the temporary files generated as part of
+            running the server.
+        """
+        if self._proc:
+            self._proc.terminate()
+        if self._outfile:
+            self._outfile.close()
+        if delete_temp_files and self.temp_dir:
+            shutil.rmtree(self.temp_dir)
+
+    def _install_templates(self, jars_dir: Path, template_dir: Path):
+        with open(_JARS_FILE) as jf:
+            template_zip_file = jars_dir.joinpath(jf.readline().strip())
+        with zipfile.ZipFile(template_zip_file) as z:
+            # should really check to see that the entries are safe, but it's our zipfile, so
+            # don't bother for now.
```

Comment:
😱

Reply:
Yeah don't do this in production at home, kids